### PR TITLE
Disable indexes with non-empty "appended" or "deleted" files if hybrid-scan is disabled

### LINKED ISSUE #193: Hyperspace should not pick indexes if hybrid scan is disabled and "appendedFiles" or "deletedFiles" in metadata is non-empty
### Describe the issue

Thanks @pirz for collaborating in identifying this issue and coming up with example scenarios

With hybrid scan disabled:
If the metadata contains some non-empty `appendedFiles` or `deletedFiles`, it is possible that based on current version of the code, the signature matches for the latest version of the index. Hyperspace will wrongly pick this index without regards to the additional `appendedFiles` or `deletedFiles` values.

Consider the following examples:
```
Example 1
Original Source files: f1, f2, f3
Index Details:
    index contents = v0(f1f2f3)  // NOTE THAT INDEX REPRESENTS THE DATA IN SOURCE
    source files = f1, f2, f3
    source sign. = "f1f2f3"

<Append f4, run some form of metadata update (e.g. refreshIndex(mode = "quick")) such that f4 becomes part of `appendedFiles` list.>
Source files: f1, f2, f3, f4
Index:
    index contents = v0(f1f2f3) // WHEN SOURCE IS UPDATED TO LATEST, INDEX GETS OUT OF SYNC
    appendedFiles = f4
    source files = f1, f2, f3, f4         // This is to maintain the semantics that metadata represents the latest snapshot of source
    source sign. = "f1f2f3f4"            // SIGNATURE IS IN SYNC WITH LATEST DATA. WE MAY MISTAKENLY PICK IT BY MATCHING SINGATURE

<Query with hybrid scan enabled => automatically handles this scenario>
<Query with hybrid scan disabled>=> 
    Expected Behavior => Skip index, fallback to data
    Current behavior => index is used. It does not contain records from f4 and should not be picked.
```

```
Example 2
Original Source files: f1, f2, f3
Index Details:
    index contents = v0(f1f2f3)  // NOTE THAT INDEX REPRESENTS THE DATA IN SOURCE
    source files = f1, f2, f3
    source sign. = "f1f2f3"

<Delete f3, run some form of metadata update (e.g. refreshIndex(mode = "quick")) such that f3 becomes part of `deletedFiles`list.>
Source files: f1, f2
Index:
    index contents = v0 (f1, f2, f3) // WHEN SOURCE IS UPDATED TO LATEST, INDEX GETS OUT OF SYNC
    deletedFiles = f3
    source files = f1, f2            // This is to maintain the semantics that metadata represents the latest snapshot of source
    source sign. = "f1f2"            // SIGNATURE IS IN SYNC WITH LATEST DATA. WE MAY MISTAKENLY PICK IT BY MATCHING SINGATURE

<Query with hybrid scan enabled => automatically handles this scenario with #171>
<Query with hybrid scan disabled>=> 
    Expected Behavior => Skip index, fallback to data
    Current behavior => index is used. It contains extra records from f3 which should not be picked.
```
Refer [here](https://github.com/microsoft/hyperspace/blob/master/src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala#L109-L111)

This can be fixed by adding additional checks that "appendedFiles" and "deletedFiles" must be empty.

### PR DESCRIPTION
**Pull request by apoorvedave1**
### What is the context for this pull request?
Fixes #193 

 - **Tracking Issue**: #193 
 - **Parent Issue**: #187

### What changes were proposed in this pull request?
Remove any index candidate for being picked when:
 - hybrid scan is disabled AND
 - either `deletedFiles` or `appendedFiles` in index metadata is non-empty

### Does this PR introduce _any_ user-facing change?
no. It's a bug fix.

### How was this patch tested?
added unit tests

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala [hunk 1]
```diff
@@ -107,7 +107,10 @@ object RuleUtils {
       indexes.filter(index =>
         index.created && isHybridScanCandidate(index, filesByRelations.flatten))
     } else {
-      indexes.filter(index => index.created && signatureValid(index))
+      indexes.filter(
+        index =>
+          index.created && signatureValid(index) &&
+            index.deletedFiles.isEmpty && index.appendedFiles.isEmpty)
```
Comment (sezruby) [inline src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: This would be ok but might cause some bad user experience:

1) refreshIncremental
2) failed to apply index
3) no idea or clue for the failure until the user checks index log entry ..

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Thanks @sezruby for bringing this up, i think that's ok. This behavior is expected. I think this experience is exactly same as the following:
1. user creates index
2. user updates source data
3. fail to apply index
4. no idea or clue for the failure until the user checks index log entry ..

Please feel free to suggest how to improve this experience. We can create an issue for the same and fix it in subsequent PRs.

On the other hand, if we don't do this, it will lead to incorrect results.

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 2]
```diff
@@ -280,6 +280,62 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
+    assert(!usableIndexes.exists(_.name.equals("t1iTest")))
+  }
+
+  test(
+    "RuleUtils.getCandidateIndexes: Verify no indexes are matched even if signature matches but " +
+      "hybrid scan is disabled and 'appendedFiles' is non-empty.") {
+    assert(!HyperspaceConf.hybridScanEnabled(spark))
+    // Here's an index where the singature is in sync with the latest data, but "deletedFiles"
+    // list is non-empty.
+    val entry =
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:317]: can you run auto format?

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 3]
```diff
@@ -280,6 +280,62 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
+    createIndex("t1iTest", Seq(t1c1), Seq(t1c3), t1ProjectNode, appendedFiles = Seq("f1"))
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:318]: if you are creating with the same name, doesn't it just ignore?

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 4]
```diff
@@ -280,6 +280,62 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
+        LogicalPlanSignatureProvider
+          .create(entry.signature.provider)
+          .signature(t1ProjectNode)
+          .get))
+
+    val indexManager = IndexCollectionManager(spark)
+    val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+    assert(allIndexes.exists(_.name.equals("t1iTest")))
+
+    val usableIndexes = RuleUtils.getCandidateIndexes(spark, allIndexes, t1ProjectNode)
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:333]: can't you just pass the index entry you created? why do we need to go thru indexmanager, etc.?

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 5]
```diff
@@ -280,6 +280,62 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
+    // Verify that even if signature matched, this index is not picked because of non-empty
+    // "deleted" files.
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:307]: To me, this doesn't really validate anything. You should create a valid index log entry, run it thru `getCandidateIndexes` and make sure it returns the index log entry. Create the same but now with `deletedFiles` and make sure nothing returns from `getCandidateIndexes`.

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 6]
```diff
@@ -280,6 +280,62 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
...
   }
 
+  test(
+    "RuleUtils.getCandidateIndexes: Verify no indexes are matched even if signature matches but " +
+      "hybrid scan is disabled and 'deletedFiles' is non-empty.") {
+    assert(!HyperspaceConf.hybridScanEnabled(spark))
+    // Here's an index where the singature is in sync with the latest data, but "deletedFiles"
+    // list is non-empty.
+    val entry =
+      createIndex("t1iTest", Seq(t1c1), Seq(t1c3), t1ProjectNode, deletedFiles = Seq("f1"))
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:290]: can we rename this function to `createIndexLogEntry`? I know it will touch few files, but I think this is a good PR to handle it.

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 1 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: The user experience currently is not good either. For example, with hybrid scan off, if you add a file, it will fail in `signatureValid` without letting the user know. We have been talking about exposing "why not" API that tells the user why an index was not picked up. Until we introduce that kind of API, I think the user experience will remain "not desirable" - meaning not good. 😄

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: I think @apoorvedave1 beat me on this comment 😄

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 2 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala]: thanks @imback82 , fixed

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 3 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala]: thanks @imback82 , I missed it. I have updated the tests anyway based on your other suggestion.

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 4 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala]: oh this is much smarter!, thanks @imback82 ,

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 5 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala]: thanks @imback82 , udpated tests

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 6 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala]: thanks @imback82 , fixed

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleTestSuite.scala [hunk 7]
```diff
@@ -43,7 +45,7 @@ trait HyperspaceRuleTestSuite extends HyperspaceSuite {
           Seq(
             Relation(
               Seq("dummy"),
-              Hdfs(Properties(Content(Directory("/")))),
+              Hdfs(Properties(Content(Directory("/")), appendedFiles, deletedFiles)),
               "schema",
               "format",
               Map())),
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleTestSuite.scala:51]: Can you update 
```
logManager.writeLog(0, indexLogEntry)
```
to
```
assert(logManager.writeLog(0, indexLogEntry))
```
in this `def`?

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala [hunk 8]
```diff
@@ -280,6 +280,19 @@ class RuleUtilsTest extends HyperspaceRuleTestSuite with SQLHelper {
     }
   }
 
+  test(
+    "RuleUtils.getCandidateIndexes: Verify indexes with non-empty 'deletedFiles' or " +
+      "'appendedFiles' are not usable indexes if hybrid scan is disabled.") {
+    assert(!HyperspaceConf.hybridScanEnabled(spark))
```
Comment (imback82) [inline src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:286]: can we just do `withSQLConf`?

**Review by imback82 — COMMENTED** [review]
Few minor comments, but LGTM, thanks @apoorvedave1!

File: src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleTestSuite.scala (hunk 7 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/HyperspaceRuleTestSuite.scala:51]: thanks @imback82 , updated

File: src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala (hunk 8 above)
Comment (apoorvedave1) [reply src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:286]: thanks @imback82 , updated

**Review by imback82 — APPROVED** [review]

**Comment by imback82** [comment]
@apoorvedave1 Can you help creating an issue to address improving the user experience when indexes are not selected (e.g., the "why not" API)? I think we should address this sooner than later (a good candidate for the next release - 0.5). Thanks!

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: @imback82 @apoorvedave1 
This is why I suggested quick failure (assert) rather than storing deletedFiles but maybe we could hybrid scan with it later. 

Now we are proposing "mutable" dataset, I think we need to address this issue with more care. 

Anyway I'm fine with the current approach as it's also valid :) 

Thanks all!

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: This check will be needed for quick mode (metadata only update) anyway, no?

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Another way to avoid the check is to change the semantics of the Source.Content.

Currently, the source files and their signature is based on the "latest" snapshot when the action was run. "appendedFiles" and "deletedFiles" are the new/deleted files "since" the index "data" was created.

We moved to this semantics while reviewing #170, where signature was based on the "latest" but source files were "since".

Another way is to make everything "since" such that:
The sources files and their signature is "when" the index "data" was created. "appendedFiles" and "deletedFiles" are the new/deleted files "since" the index "data" was created.

Can we discuss pros/cons for each approach?

(For example, are we losing anything if we move the current approach (mix of "latest" and "since") to the new approach (everything "since")? I think the current approach is to favor hybrid scan such that if the signature matches, we can just utilize "appendedFiles" and "deletedFiles" - which is not yet implemented btw.)

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: In my perspective, I prefer everything "since" for the incremental mode - and not allow deleted or appended files in the metadata like full mode. But for this we need to merge RefreshAppend and RefreshDelete. 

For quick mode, we could use the current approach - update signature with the latest and collect appended and deleted files. 

This is what I was thinking first and I think we could avoid possible complex cases and conditions with this approach.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Can we first agree on what we want to store in the metadata? How we handle the metadata in different refresh modes is an orthogonal discussion. (For example, we should store the same info (everything "since" or mix) regardless of refresh modes).

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: > This check will be needed for quick mode (metadata only update) anyway, no?

for continuing with this discussion, I don't think we need this check for metadtata only update either. Could you please give me a scenario?
I was thinking metadata only update is much simpler. with `RefreshMetadataAction` (or something I read in another PR), we can just update the 'deleted' and 'appended' files in metadata. Isn't that enough? Why do we need the assert(deleted.nonEmpty) I don't really understand.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: I meant "This check" = ` index.deletedFiles.isEmpty && index.appendedFiles.isEmpty)` in this file, not the assert. What I mean is that with the current semantics, we need this check regardless of the refresh modes (which are orthogonal).

Comment (apoorvedave1) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: yeah these checks are required with the current semantics regardless of the refresh mode. So here's my understanding:

If hybrid scan is enabled, we know the data-source-changes (either adds or deletes) will be handled by hybrid scan.
If it is not, the right way to handle this is DO NOT USE the index because it's not up to date with the data.

So I am currently fine with this semantic.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Sorry I misread. I thought "since" means "latest". So "latest"-`source.Content` indicates the source files of the index data in that version. Then I think "mixed" approach is better as there is no special use of "since" "appendedFiles" and "deletedFiles" during query time or refresh action?

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Oh OK. The use of "latest" could be confusing. Just to make sure everyone is on the same page, let's assume the following events in sequence:
- There are source files: f1, f2, f3
- An index is created.
- A source file is added: f4
- A source file is deleted: f1
- Refresh "quick" (metadata only operation)

With the current semantics (mix of "latest" and "since"), we will have the following in the latest index log entry:
- source files / signature: f2, f3, f4 / sig(f2, f3, f4)
- appendedFiles: f4 
- deletedFiles: f1

With the semantics of always using "since":
- source files / signature: f1, f2, f3 / sig(f1, f2, f3)
- appendedFiles: f4 
- deletedFiles: f1

And I am hearing that the current semantics (mix of "latest" and "since") will be a better use. Is that correct?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: That's correct. The benefit of `quick` is skipping the comparison of file list for hybrid scan and using the signature instead. ( + and it also allows selective hybrid scan - disable hybrid scan config and do quick refresh only for the indexes which are candidates of hybrid scan op. )
On the other hand, with always using "since" approach, we need to check if the source file list in the given plan could be the candidate of each index in some how at query execution time.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: What if we capture the latest source signature as separate field? For example, we introduce `Update` case class that captures "appendedFiles", "deletedFiles", and "latestSourceFilesSignature"' for the `Source.Content`? (I am just exploring the best option for both semantics / usage before the next release, so please bear with me 😄 ). It seems that if the signature doesn't match, we need to fallback to iterating files for hybrid scan anyway.


> ( + and it also allows selective hybrid scan - disable hybrid scan config and do quick refresh only for the indexes which are candidates of hybrid scan op. )

Can you also elaborate on this scenario? I didn't get the part where you disable the hybrid scan config and do quick refresh since the paths are different (one for query and another for non-query).

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: > What if we capture the latest source signature as separate field? For example, we introduce Update case class that captures "appendedFiles", "deletedFiles", and "latestSourceFilesSignature"' for the Source.Content? 

This could be an option though the original signature value won't be used anymore.
But in this way, we could keep the same semantic for Source.Content. 
And in the case of "full" and "incremental" refresh, `Source.Content` contains the source files at refresh (both full and incremental) the same as now?

> Can you also elaborate on this scenario? I didn't get the part where you disable the hybrid scan config and do quick refresh since the paths are different (one for query and another for non-query).

For example, there're 5 indexes but one of them has many "diff" files in its source files since the index build. But a user want to use hybrid scan for an index with very little diff files. If global hybrid scan config is enabled, Hyperspace tries to apply hybrid scan for all candidate indexes - if there's no proper hybrid scan threshold in candidate selection. 
So `quick` refresh can be an option for this situation:
- disable hybrid scan config
- quick refresh for some indexes with little diff
- hybrid scan is applied only for `quick` refreshed indexes

Although this is different context, I just wanted to describe the possible use case of `quick` refresh.

Comment (rapoth) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: @imback82 @sezruby Should we move this discussion to an issue so we can follow-up properly through a PR, if necessary? If you agree, @sezruby - can you please create an issue with all the approaches + pros/cons so everyone is on the same page? You can paste the link to that issue here as a final reply.

**Comment by rapoth** [comment]
> @apoorvedave1 Can you help creating an issue to address improving the user experience when indexes are not selected (e.g., the "why not" API)? I think we should address this sooner than later (a good candidate for the next release - 0.5). Thanks!

@apoorvedave1 Can you take care of creating this issue with the relevant details?

File: src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala (hunk 1 above)
Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: > This could be an option though the original signature value won't be used anymore.

Yes, it will be used only if `Update` is `isEmpty == true` (assuming we can capture `Update` as `Option[]`).

> But in this way, we could keep the same semantic for Source.Content.

Yes, I think the code logic will be simpler as well.

> And in the case of "full" and "incremental" refresh, `Source.Content` contains the source files at refresh (both full and incremental) the same as now?

Well, it will store the source files that are indexed, so it could be different if there are deleted files before `RefreshAppend` runs.

> Although this is different context, I just wanted to describe the possible use case of `quick` refresh.

Oh, you mean refresh "on-the-fly" during query time.

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Created #198 to track this issue. 
BTW
>>Although this is different context, I just wanted to describe the possible use case of quick refresh.
>
>Oh, you mean refresh "on-the-fly" during query time.

it's possible but I meant a user can choose the candidate indexes for hybrid scan - with disabled global config.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: OK, in this case, I still don't understand 😄. @rapoth do you understand this scenario?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Ok let me try again 😁 
1. disable "**global**" hybrid scan config
2. user can choose some "beneficial - with high perf improvement" indexes with little diff and do `quick` refresh for the indexes
3.  hybrid scan can be performed for `quick` refreshed indexes even if the global config is disabled. The other unrefreshed indexes with diff won't be the candidates for hybrid scan.

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Thanks 👍 . So in a nutshell, the selective hybrid scan is "we can apply the hybrid scan even if the user didn't enable it only when the signature matches", right?

Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Right. That's the point 👍

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:113]: Thanks!
